pandajedi/jedicore/JediDBProxy.py

```diff
@@ -13456,23 +13456,41 @@ def updateInputFilesStagedAboutIdds_JEDI(self, jeditaskid, scope, filenames):
             varMap[':new_status'] = 'pending'
             resGD = self.cur.fetchall()
             primaryID = None
+            params_key_list = []
+            var_map_datasetids = {}
             if len(resGD) > 0:
-                for idx, (id, masterID) in enumerate(resGD):
-                    key = ':datasetID_{0}'.format(idx)
-                    sqlUF += '{0},'.format(key)
-                    varMap[key] = id
+                for idx, (tmp_datasetID, masterID) in enumerate(resGD):
                     if masterID is None:
-                        primaryID = id
-                sqlUF = sqlUF[:-1]
-                sqlUF += ') '
+                        primaryID = tmp_datasetID
+                    key = ':datasetID_{0}'.format(idx)
+                    params_key_list.append(key)
+                    var_map_datasetids[key] = tmp_datasetID
+            else:
+                to_update_files = False
+            # update files
+            if to_update_files:
                 # loop over filenames
                 varMaps = []
-                for filename in filenames:
+                for filename, (datasetid, fileid) in filenames_dict.items():
                     tmp_varMap = varMap.copy()
                     if scope != 'pseudo_dataset':
                         tmp_varMap[':lfn'] = filename
                     else:
                         tmp_varMap[':lfn'] = '%' + filename
+                    if datasetid is not None:
+                        # with datasetID from message
+                        sqlUF += 'AND datasetID=:datasetID '
+                        tmp_varMap[':datasetID'] = datasetid
+                    else:
+                        # without datasetID from message
+                        params_key_str = ','.join(params_key_list)
+                        datesetid_list_str = f'AND datasetID IN ({params_key_str}) '
+                        sqlUF += datesetid_list_str
+                        tmp_varMap.update(var_map_datasetids)
+                    if fileid is not None:
+                        # with fileID from message
+                        sqlUF += 'AND fileID=:fileID '
```

Comment:
Doesn't this add many "AND fileID=:fileID " in sqlUF? It would be better to add a new parameter in the updateInputFilesStagedAboutIdds_JEDI method like file_data_with_ids=False or something, so that you wouldn't have to check the presence of IDs for each file.

Reply:
Thanks. That was definitely a bug.
I fixed it in a different way - The challenge is it is not guaranteed that files in the same message from iDDS all have or don't have fileID (some do, the others don't); meanwhile in updateInputFilesStagedAboutIdds_JEDI it updates DB with executemany so it requires to split into multiple executemanys with different SQL according to the existence of fileID or datasetID